metricflow/specs/where_filter_transform.py

```diff
@@ -42,91 +31,29 @@ def __init__(  # noqa: D
-                entity_path=(
-                    tuple(EntityReference(element_name=arg) for arg in entity_path) + structured_name.entity_links
-                ),
-            )
-            assert call_parameter_set in call_parameter_sets.dimension_call_parameter_sets
-
-            dimension_spec = WhereSpecFactory._convert_to_dimension_spec(call_parameter_set)
-            dimension_specs.append(dimension_spec)
-            return self._column_association_resolver.resolve_spec(dimension_spec).column_name
-
-        def _time_dimension_call(
-            time_dimension_name: str, time_granularity_name: str, entity_path: Sequence[str] = ()
-        ) -> str:
-            """Gets called by Jinja when rendering {{ TimeDimension(...) }}."""
-            structured_name = DunderedNameFormatter.parse_name(time_dimension_name)
-            call_parameter_set = TimeDimensionCallParameterSet(
-                time_dimension_reference=TimeDimensionReference(element_name=structured_name.element_name),
-                time_granularity=TimeGranularity(time_granularity_name),
-                entity_path=(
-                    tuple(EntityReference(element_name=arg) for arg in entity_path) + structured_name.entity_links
-                ),
-            )
-            assert call_parameter_set in call_parameter_sets.time_dimension_call_parameter_sets
-
-            time_dimension_spec = WhereSpecFactory._convert_to_time_dimension_spec(call_parameter_set)
-            time_dimension_specs.append(time_dimension_spec)
-            return self._column_association_resolver.resolve_spec(time_dimension_spec).column_name
-
-        def _entity_call(entity_name: str, entity_path: Sequence[str] = ()) -> str:
-            """Gets called by Jinja when rendering {{ Entity(...) }}."""
-            structured_name = DunderedNameFormatter.parse_name(entity_name)
-            call_parameter_set = EntityCallParameterSet(
-                entity_reference=EntityReference(element_name=entity_name),
-                entity_path=(
-                    tuple(EntityReference(element_name=arg) for arg in entity_path) + structured_name.entity_links
-                ),
-            )
-            assert call_parameter_set in call_parameter_sets.entity_call_parameter_sets
-
-            entity_spec = WhereSpecFactory._convert_to_entity_spec(call_parameter_set)
-            entity_specs.append(entity_spec)
-            return self._column_association_resolver.resolve_spec(entity_spec).column_name
-
         try:
             rendered_sql_template = jinja2.Template(
                 where_filter.where_sql_template, undefined=jinja2.StrictUndefined
             ).render(
-                Dimension=_dimension_call,
-                TimeDimension=_time_dimension_call,
-                Entity=_entity_call,
+                {
+                    "Dimension": WhereFilterDimensionFactory(
+                        call_parameter_sets, dimension_specs, self._column_association_resolver
```

Comment:
The implicit update by reference on dimension_specs is challenging, because it's really not obvious that it's happening here.

I haven't come up with a good way of separating it from create, since we kind of need to do it inside of that call. I think that's why Paul originally had everything inlined above this point.

In the meantime, we could just rename the parameter to make it more obvious that the factory invocation will populate the spec list based on user input.

Reply:
Ya, that is a good point that I hadn't considered. I think it would be better if they just weren't passed in and instead just properties on `WhereFilterDimensionFactory`.

Reply:
I made that update [here](https://github.com/dbt-labs/metricflow/pull/717/commits/19c79cd80c94d1956540e963f789afbc64e898b3). Let me know if you would prefer a different solution.

Reply:
We talked in person about this.